Several fixes for UEB back-translation

jcsteh opened the pull request:
This includes the following:
- Don't backtranslate the dot 4, dot 5 and dot 5 6 indicators as punctuation when not followed by another cell. These indicators prefix various contractions and symbols, but should not have any meaning by themselves. This is needed to improve cell by cell backtranslation of input from a braille keyboard. See below for further explanation.
- Exclamation mark at the end of a word no longer produces "ff".
  - Fixes #212.
- Question mark at the end of a word no longer produces "his".
- Open and close quotes no longer produce "his" and "was".
- Colon at the end of a word no longer produces "cc".
- Semi-colon at the end of a word no longer produces "bb".
- The gg contraction in the middle of a word no longer produces "were".

Regarding the indicators:

These indicators prefix various contractions and symbols, but should not have any meaning by themselves. This is needed to improve cell by cell backtranslation of input from a braille keyboard.

See also #210, which is needed to make this change work as intended.

When doing uncontracted translation, NVDA sends characters to the system as soon as they are produced. So, for example, if you type dots 3 4 5 6 - 1, the 3 4 5 6 doesn't produce text, but when you press the dot 1, NVDA sends the character "1" to the system. If you then type dots 1 2, we send "2" to the system. However, this relies on liblouis not producing any text for cells which require an additional cell to be valid. This works for number sign, capital sign, etc. However, it fails for indicators which aren't specifically defined; i.e. those that prefix symbols or contractions. An example is dot 4 used to type the "@" symbol.

Similarly, when using contracted braille, we use a similar technique to report a "guess" to the user of the characters they're entering. Again, the number example works, but this breaks when trying to type "name" because the dot 5 produces the quote symbol ("). The "tion" contraction has a similar problem: when you type dots 5 6, liblouis produces a semi-colon (;).

Ideally, we'd do something similar for English U.S. and U.K. grade 1/2 (and probably many other uncontracted/contracted tables). However, it's a bit trickier for those cases because they use chardefs.cti and changing that has much wider implications. It could break tables where these aren't indicators and actually should backtranslate to punctuation. To solve this, we'd need either:
1. An alternative approach where we have a new opcode to define indicators as such in the relevant uncontracted/contracted tables; or
2. Separation of these symbols that could be used as indicators out of the character definition tables. This could get very messy, though.

See also [this mailing list thread](https://www.freelists.org/post/liblouis-liblouisxml/ueb-backtranslation-numbers,2).

egli commented:
Looks fine to me but I use neither ueb nor backtranslation, so you might want to have a quick chat with @MikeGray-APH to see what they think.

jcsteh commented:
I've extended this PR to cover several other UEB back-translation issues, again because of the conflicts that would result if I did these separately. I've also added tests for all of these. See the updated PR description for the issues that are now fixed.

MikeGray-APH commented:
This may run into an issue we are having about when a cell should be considered undefined.  I started the discussion on the mail list.

MikeGray-APH commented:
I will pull this into my branch at https://github.com/MikeGray-APH/liblouis before I do a pull request into master, unless anyone has any objections?

egli commented:
@MikeGray-APH I presume that you looked at this and that this could go in. Correct?

bertfrees commented:
See discussion on ML: http://www.freelists.org/post/liblouis-liblouisxml/Undefined-Cells

egli commented:
@jcsteh can you please have a look. I must have botched something up when merging. Two test fail now
```

FAIL: yaml/en-ueb-03-symbols
============================

Input:    '151° 12' 40" E'
Expected: '#aea~j #ab' #dj,7 ;,e' (length 21)
Received: '#aea~j #ab' #dj" ;,e' (length 20)
Diff: Expected ',' but received '"' in index 15
lou_checkyaml:./yaml/en-ueb-03-symbols.yaml:193: Failure
FAILURE (85 tests, 1 failure)
FAIL yaml/en-ueb-03-symbols.yaml (exit status: 1)

FAIL: yaml/en-ueb-g2_backward
=============================

Input:    '⠁⠖⠁'
Expected: 'affa' (length 4)
Received: 'a!a' (length 3)
Diff: Expected 'f' but received '!' in index 1
lou_checkyaml:./yaml/en-ueb-g2_backward.yaml:6: Failure
Input:    '⠦'
Expected: 'his' (length 3)
Received: '?' (length 1)
Diff: Expected 'h' but received '?' in index 0
lou_checkyaml:./yaml/en-ueb-g2_backward.yaml:7: Failure
Input:    '⠦⠁'
Expected: '"a' (length 2)
Received: '?a' (length 2)
Diff: Expected '"' but received '?' in index 0
lou_checkyaml:./yaml/en-ueb-g2_backward.yaml:9: Failure
Input:    '⠴'
Expected: 'was' (length 3)
Received: '"' (length 1)
Diff: Expected 'w' but received '"' in index 0
lou_checkyaml:./yaml/en-ueb-g2_backward.yaml:10: Failure
Input:    '⠦⠁⠴'
Expected: '"a"' (length 3)
Received: '?a"' (length 3)
Diff: Expected '"' but received '?' in index 0
lou_checkyaml:./yaml/en-ueb-g2_backward.yaml:12: Failure
Input:    '⠶'
Expected: 'were' (length 4)
Received: '′' (length 1)
Diff: Expected 'w' but received '′' in index 0
lou_checkyaml:./yaml/en-ueb-g2_backward.yaml:13: Failure
Input:    '⠁⠶⠁'
Expected: 'agga' (length 4)
Received: 'a′a' (length 3)
Diff: Expected 'g' but received '′' in index 1
lou_checkyaml:./yaml/en-ueb-g2_backward.yaml:14: Failure
Input:    '⠆'
Expected: 'be' (length 2)
Received: ';' (length 1)
Diff: Expected 'b' but received ';' in index 0
lou_checkyaml:./yaml/en-ueb-g2_backward.yaml:15: Failure
Input:    '⠆⠁'
Expected: 'bea' (length 3)
Received: ';a' (length 2)
Diff: Expected 'b' but received ';' in index 0
lou_checkyaml:./yaml/en-ueb-g2_backward.yaml:16: Failure
Input:    '⠁⠆⠁'
Expected: 'abba' (length 4)
Received: 'a;a' (length 3)
Diff: Expected 'b' but received ';' in index 1
lou_checkyaml:./yaml/en-ueb-g2_backward.yaml:17: Failure
Input:    '⠒⠁'
Expected: 'cona' (length 4)
Received: ':a' (length 2)
Diff: Expected 'c' but received ':' in index 0
lou_checkyaml:./yaml/en-ueb-g2_backward.yaml:20: Failure
Input:    '⠁⠒⠁'
Expected: 'acca' (length 4)
Received: 'a:a' (length 3)
Diff: Expected 'c' but received ':' in index 1
lou_checkyaml:./yaml/en-ueb-g2_backward.yaml:22: Failure
FAILURE (23 tests, 12 failures)
FAIL yaml/en-ueb-g2_backward.yaml (exit status: 1)
```

My merged stuff is in https://github.com/liblouis/liblouis/tree/nvaccess-nobackIndicators

jcsteh commented:
I merged this myself and pushed. All tests now pass.
